GameEngine/Deck.cs

```diff
@@ -1,84 +1,27 @@
+        public abstract IDeck Clone();
+        public abstract CardType[] Draw(int numberDesired);
+        public abstract CardType[] Sample(int numberDesired);
+        public virtual CardType[] SampleAll() => Sample(Count);
+        public override abstract bool Equals(object o);
+        public override abstract int GetHashCode();
 
-        private Deck() { }
-
-        public Deck(int gameSizeMultiplier, int? numberOfSunCards = null)
-        {
-            var defaultSunCards = gameSizeMultiplier + 4 * gameSizeMultiplier / 3;
-            BuildDeck(gameSizeMultiplier, numberOfSunCards ?? defaultSunCards);
-            Shuffle();
-        }
-
-        public CardType[] Draw(int numberDesired)
-        {
-            var numberToDraw = Math.Min(numberDesired, Cards.Count);
-            var cards = Cards.GetRange(0, numberToDraw);
-            Cards.RemoveRange(0, numberToDraw);
-            return cards.ToArray();
-        }
-
-        private void BuildDeck(int gameSizeMultiplier, int sunCards)
-        {
-            Cards = new List<CardType>(Enumerable.Repeat(0, gameSizeMultiplier)
-                .SelectMany(cards => CardTypeExtensions.OneCardOfEachColor)
-                .Concat(Enumerable.Repeat(CardType.Sun, sunCards)));
-        }
-
-        private void Shuffle()
-        {
-            // http://en.wikipedia.org/wiki/Fisher-Yates_shuffle
-            int n = Cards.Count;
-            while (n > 1)
-            {
-                n--;
-                int k = SeededRandom.Next(n + 1);
-                CardType value = Cards[k];
-                Cards[k] = Cards[n];
-                Cards[n] = value;
-            }
-        }
-
-        public override bool Equals(object o)
-        {
-            var other = o as Deck;
-            return other != null
-                && Cards.SequenceEqual(other.Cards);
-        }
-
-        public override int GetHashCode()
-        {
-            var hashCode = 1217012356;
-            unchecked
-            {
-                hashCode += EqualityComparer<List<CardType>>.Default.GetHashCode(Cards);
-            }
-            return hashCode;
-        }
-
-        public static bool operator ==(Deck left, Deck right)
-        {
-            return EqualityComparer<Deck>.Default.Equals(left, right);
-        }
-
-        public static bool operator !=(Deck left, Deck right)
+        protected int StartingCount(int multiplier, int? requestedSunCards = null)
         {
-            return !(left == right);
+            return multiplier * CardTypeExtensions.OneCardOfEachColor.Length
+                + StartingSunCount(multiplier, requestedSunCards);
         }
 
-        public Deck Clone()
+        protected int StartingSunCount(int multiplier, int? requestedSunCards = null)
         {
-            return new Deck()
-            {
-                Cards = new List<CardType>(Cards)
-            };
+            return requestedSunCards ?? 7 * multiplier / 3;
```

Comment:
💡 Could we sweep the game constants (7 and 3) into some interface defined constants?

Reply:
I'm not sure what I'd name them. And I'd prefer not to save off the ratio since I'm leveraging integer division. I recognize that they're magic numbers and that magic is bad. I just don't know what name(s) would make them seem less magical.